### DISCUSSION TITLE
progress bar with other infolabels

### DISCUSSION THREAD

**Discussion by nico1080** [mattblovell/kodi_panel#40] [Ideas]
Currently the progress bars use `VideoPlayer.Time` and `VideoPlayer.Duration` for the video section

However for live TV these infolabel does not work. (at least in my setup with tvheadend and kodi 18.9)
The infolabel `PVR.EpgEventElapsedTime` and `PVR.EpgEventDuration` send back the correct infomation.

How can I use theses infolabel for progress bar only in my live TV layout, and keep the previous infolabels for my other layout?
Maybe with the `ELEMENT_CB` it will work? But I am not sur how it work.
Or maybe override the `prog `function ? similary to what I have done with the `calc_progress `function

**Comment by mattblovell** [comment]
With the callback framework that's in place, this should already be possible.

The ``progress_bar()`` function is used by both audio and video screens.  It is the calling functions that are responsible for invoking ``calc_progress()``, passing it either the ``MusicPlayer`` or ``VideoPlayer`` time and duration InfoLabels.

So, mimicking the structure of that code, you can get ``progress_bar()`` invoked with whatever calculated progress is appropriate.  The setup would be roughly as follows...

First, ensure the ``PVR.EpgEventElapseTime`` and ``PVR.EpgEventDuration`` have been added to the ``VIDEO_LABELS`` list in the setup.toml file.

In your script, define a new ``element_*`` callback function:
```
def element_livetv_progbar(image, draw, info, field, screen_mode, layout_name):
    # Calculate progress in media, using InfoLabels appropriate for LiveTV
    prog = kodi_panel_display.calc_progress(
        info["PVR.EpgEventElapsedTime"],
        info["PVR.EpgEventDuration"],
        kodi_panel_display.video_dmode.name
    )    

   show_prog = True

   # If the field has a display conditional
   # defined, let's test that to decide if we should proceed.
   if ("display_if" in field or
       "display_ifnot" in field):
       show_prog = kodi_panel_display.check_display_expr(field,
                                      info,
                                      kodi_panel_display.ScreenMode.AUDIO,
                                      kodi_panel_display.video_dmode.name)

   if show_prog:
       kodi_panel_display.progress_bar(
           draw, field, prog,
           use_long_len = (info['PVR.EpgEventDuration'].count(":") == 2)
       )

```

The above function takes excerpts from ``video_screens()`` and ``video_screen_dynamic()``.

Add the above callback to the ``ELEMENT_CB`` dictionary in the now-usual fashion, naming it whatever you would like (e.g., ``livetv_prog``).

In LiveTV layouts, do *not* have a ``prog`` element within the top-level of that layout's dictionary.  Instead, take all of the key/value pairs that you would have used for the progress bar and use them for a new entry within the layouts ``fields`` array.  It must be named such that a callback occurs to the above function.  All of the key/value pairs should end up getting passed through the callback, down to the existing ``progress_bar()`` function, which will then draw the bar as usual, using whatever mechanism you specified to calculate the fractional media progress.

I typed the above into github's discussion widget directly, so I'm not sure all of the spaces are entirely consistent, etc.  Hopefully, though, you get the main direction I am trying to convey.

Cheers,
  Matt

**Comment by nico1080** [comment]
Works great, I put it in my fork: dd28c59eda17341770f75ee2ac38d8cb6e0d6478
